aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/decorators/AwsCombinedCodegenDecorator.kt

```diff
@@ -0,0 +1,102 @@
+import software.amazon.smithy.rust.codegen.core.rustlang.rustBlockTemplate
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeConfig
+import software.amazon.smithy.rust.codegen.core.smithy.RustCrate
+import software.amazon.smithy.rust.codegen.core.smithy.customize.writeCustomizations
+import software.amazon.smithy.rustsdk.AwsCustomization
+import software.amazon.smithy.rustsdk.AwsSection
+import software.amazon.smithy.rustsdk.awsTypes
+import software.amazon.smithy.rustsdk.servicedecorators.apigateway.ApiGatewayDecorator
+import software.amazon.smithy.rustsdk.servicedecorators.auth.DisabledAuthDecorator
+import software.amazon.smithy.rustsdk.servicedecorators.ec2.Ec2Decorator
+import software.amazon.smithy.rustsdk.servicedecorators.glacier.GlacierDecorator
+import software.amazon.smithy.rustsdk.servicedecorators.route53.Route53Decorator
+import software.amazon.smithy.rustsdk.servicedecorators.s3.S3Decorator
+import software.amazon.smithy.rustsdk.servicedecorators.sts.STSDecorator
+
+val DECORATORS = listOf(
+    // General AWS Decorators
+    AwsEndpointDecorator(),
+    AwsFluentClientDecorator(),
+    AwsPresigningDecorator(),
+    AwsReadmeDecorator(),
+    CrateLicenseDecorator(),
+    CredentialsProviderDecorator(),
+    HttpRequestChecksumDecorator(),
+    HttpResponseChecksumDecorator(),
+    IntegrationTestDecorator(),
+    RegionDecorator(),
+    ResiliencyDecorator(),
+    SdkConfigDecorator(),
+    ServiceConfigDecorator(),
+    SigV4SigningDecorator(),
+    UserAgentDecorator(),
+
+    // Service specific decorators
+    ApiGatewayDecorator(),
+    DisabledAuthDecorator(),
+    Ec2Decorator(),
+    GlacierDecorator(),
+    Route53Decorator(),
+    S3Decorator(),
+    STSDecorator(),
+
+    // Only build docs-rs for linux to reduce load on docs.rs
+    DocsRsMetadataDecorator(DocsRsMetadataSettings(targets = listOf("x86_64-unknown-linux-gnu"), allFeatures = true)),
+)
+
+interface AwsCodegenDecorator : RustCodegenDecorator<ClientProtocolGenerator, ClientCodegenContext> {
+    fun awsCustomizations(
+        codegenContext: ClientCodegenContext,
+        baseCustomizations: List<AwsCustomization>,
+    ): List<AwsCustomization> {
+        return baseCustomizations
+    }
+}
+
+class AwsCombinedCodegenDecorator :
+    CombinedCodegenDecorator<ClientProtocolGenerator, ClientCodegenContext>(DECORATORS) {
+    override val order: Byte = -1
+
+    override fun extras(codegenContext: ClientCodegenContext, rustCrate: RustCrate) {
+        super.extras(codegenContext, rustCrate)
+
+        val awsCustomizations = orderedDecorators.filterIsInstance<AwsCodegenDecorator>()
+            .foldRight(listOf<AwsCustomization>()) { decorator, customizations ->
+                decorator.awsCustomizations(codegenContext, customizations)
+            }
+        generateImplFromRefSdkConfigForConfigBuilder(codegenContext.runtimeConfig, rustCrate, awsCustomizations)
+    }
+}
+
+fun generateImplFromRefSdkConfigForConfigBuilder(
```

Comment:
I think no because I feel that all the section writing code should be defined in one place. The combined decorator is also the place where we have access to the full list of decorators that modify an `AwsSection`.

Reply:
Organizationally, this isn't going to work out well since people will think to look in `SdkConfigDecorator` for this code rather than in `AwsCombinedCodegenDecorator`. The full list of decorators can be passed into the customizations if it's required (and I think we do that in `codegen-core`/`codegen-client`.

Reply:
Russell is adding the ability to do this as part of #1953, I can take the change from that or wait for the merge.